opm/common/utility/ThreadSafeMapBuilder.hpp

```diff
@@ -0,0 +1,158 @@
+/*
+  This file is part of the Open Porous Media project (OPM).
+
+  Copyright 2025 SINTEF Digital
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+#ifndef OPM_THREAD_SAFE_MAP_BUILDER_HPP
+#define OPM_THREAD_SAFE_MAP_BUILDER_HPP
+
+#include <algorithm>
+#include <cstddef>
+#include <list>
+#include <map>
+#include <numeric>
+#include <vector>
+
+#if _OPENMP
+#include <omp.h>
+#endif
+
+namespace Opm {
+
+//! \brief Utility class for adding multi-threading to loops that
+//!        are building maps.
+//! details Intended to be used for building maps in OpenMP loops.
+//!         In the parallel section the values are assembled into a
+//!         list per thread. After the parallel section the lists are
+//!         assembled to the map, either through RAII in the dtor,
+//!         or through a manual call to finalize().
+//!         Implements both emplace and insert_or_assign, but there are
+//!         a need for an assumption. In particular, the assumption is
+//!         that iteration loops are ordered such that the first thread
+//!         processes the first part of the range. If the order of processing
+//!         is unspecified the final map results is not guaranteed to match
+//!         for differing number of threads.
+
+//! Available insertion modes.
+enum class MapBuilderInsertionMode {
+    Emplace,          // Use first inserted value encountered
+    Insert_Or_Assign  // Use last inserted value encountered (last thread wins)
+};
+
+template<class Map>
+class ThreadSafeMapBuilder
+{
+    //! \brief Predicate for maps
+    template<class T>
+    struct is_ordered
+    {
+        constexpr static bool value = false;
+    };
+
+    template<class Key, class T, class Compare, class Allocator>
+    struct is_ordered<std::map<Key,T,Compare,Allocator>>
+    {
+        constexpr static bool value = true;
+    };
+
+public:
+    using Key = typename Map::key_type;
+    using Value = typename Map::mapped_type;
+
+    //! \brief Constructor for a given map and number of threads.
+    ThreadSafeMapBuilder(Map& map, const int num_threads,
+                         const MapBuilderInsertionMode mode = MapBuilderInsertionMode::Insert_Or_Assign)
```

Comment:
I do not like default arguments, and think this would be better without it.

Reply:
okay